File: JavaScript/src/electric_rose.js
function Battery(type, dry, charge) {
  this.type = type;
  this.dry = dry;
  this.charge = charge;
}

function ElectricRose(batteries) {
  this.batteries = batteries;
}
ElectricRose.prototype.updateCharge = function() {
  for (var i = 0; i < this.batteries.length; i++) {
    if (this.batteries[i].charge >= 0.1) {
      this.batteries[i].charge = Math.max(0, this.batteries[i].charge - 0.1);
    }

    if (this.batteries[i].dry) {
      if (this.batteries[i].charge >= 0.1) {
        this.batteries[i].charge = Math.max(0, this.batteries[i].charge - 0.1);
      }
    }
  }
}

module.exports = {
  Battery,
  ElectricRose
}

File: C/ElectricRose.c
#include <string.h>
#include <stdio.h>
#include "ElectricRose.h"
#define MAX(x, y) (((x) > (y)) ? (x) : (y))

void init_battery(Battery* battery, const char *type, bool dry, float charge)
{
    battery->dry = dry;
    battery->charge = charge;
    battery->type = strdup(type);
}

int print_battery(Battery *battery)
{
    return printf("%s, %d, %.1f\n", battery->type, battery->dry, battery->charge);
}

void update_charge(Battery batteries[], int size)
{
    int i;
    for (i = 0; i < size; i++)
    {
        if (batteries[i].charge >= 0.1) {
            batteries[i].charge = MAX(0, batteries[i].charge - 0.1);
        }

        if (batteries[i].dry) {
            if (batteries[i].charge >= 0.1) {
                batteries[i].charge = MAX(0, batteries[i].charge - 0.1);
            }
        }
    }
}
